casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala

```diff
@@ -91,32 +127,132 @@ object EstimatorHelper {
-        .toSet
+        val commEvents = comms
+          .flatMap(TuplespaceEvent.from(_, produces))
 
-      ancestorCommChannels = commEvents
-        .filterNot(isVolatile(_, consumeEvents, produceEvents))
-        .flatMap { comm =>
-          comm.consume.channelsHashes ++ comm.produces.map(_.channelsHash)
+        (produceEvents
+          .combine(consumeEvents)
+          .combine(commEvents))
+          .groupBy(_._1)
+          .mapValues[Set[TuplespaceEvent]](_.map(_._2))
+    }
+
+  final case class TuplespaceEvent(
+      incoming: TuplespaceOperation,
+      matched: Option[TuplespaceOperation]
+  )
+  final case class TuplespaceOperation(
+      polarity: Polarity,
+      cardinality: Cardinality,
+      eventHash: Blake2b256Hash
+  )
+
+  trait Polarity
+  case object Send    extends Polarity
+  case object Receive extends Polarity
+
+  trait Cardinality
+  case object Linear    extends Cardinality
+  case object NonLinear extends Cardinality
+
+  object TuplespaceEvent {
+
+    implicit private[this] def liftProduce(produce: Produce): TuplespaceOperation =
+      TuplespaceOperation(Send, if (produce.persistent) NonLinear else Linear, produce.hash)
+
+    implicit private[this] def liftConsume(consume: Consume): TuplespaceOperation =
+      TuplespaceOperation(Receive, if (consume.persistent) NonLinear else Linear, consume.hash)
+
+    def from(produce: Produce): (Blake2b256Hash, TuplespaceEvent) =
+      produce.channelsHash -> TuplespaceEvent(
+        produce,
+        None
+      )
+
+    def from(consume: Consume): Option[(Blake2b256Hash, TuplespaceEvent)] = consume match {
+      case Consume(singleChannelHash :: Nil, _, _, _) =>
+        Some(
+          singleChannelHash -> TuplespaceEvent(
+            consume,
+            None
+          )
+        )
+      case _ => None
+    }
+
+    def from(
+        comm: COMM,
+        produces: Set[Produce]
+    ): Option[(Blake2b256Hash, TuplespaceEvent)] =
+      comm match {
+        case COMM(consume, produce :: Nil, _) => {
+          val incoming: TuplespaceOperation =
+            if (produces.contains(produce)) produce else consume
+          val matched: Option[TuplespaceOperation] = Some(
+            if (incoming == liftProduce(produce)) consume else produce
+          )
+          Some(
+            produce.channelsHash -> TuplespaceEvent(
+              incoming,
+              matched
+            )
+          )
         }
-        .toSet
+        case _ => None
+      }
+  }
+
+  implicit class TuplespaceEventOps(val ev: TuplespaceEvent) extends AnyVal {
 
-      ancestorChannels = ancestorProduceChannels ++ ancestorConsumeChannels ++ ancestorCommChannels
-      res              = ancestorChannels.map(_.toByteString)
-    } yield res
+    private[casper] def conflicts(other: TuplespaceEvent): Boolean =
+      if (ev.incoming.polarity == other.incoming.polarity)
+        (for {
+          a <- ev.matched
+          b <- other.matched
```

Comment:
are you happy with the `a` `b` names?

Reply:
I am, but I could use better names for readability's sake. What would you think about `leftMatched` and `rightMatched`?

Reply:
thisMatched, otherMatched? there is no left/right in this context imo